lib/srv/db/cloud/iam.go

```diff
@@ -125,3 +192,168 @@ func (c *IAM) getAWSIdentity(ctx context.Context) (aws.Identity, error) {
+		}
+	}()
+
+	if task.isSetup {
+		return configurator.setupIAM(ctx)
+	}
+	return configurator.teardownIAM(ctx)
+}
+
+// addTask add a task for processing.
+func (c *IAM) addTask(task iamTask) error {
+	select {
+	case c.tasks <- task:
+		return nil
+
+	default:
+		return trace.LimitExceeded("failed to create IAM task for %v", task.database.GetName())
+	}
+}
+
+// deleteOldPolicy removes old inline policies "teleport-<host-id>" for the
+// caller identity. DELETE IN 11.0.
+func (c *IAM) deleteOldPolicy(ctx context.Context) {
+	oldPolicyName := "teleport-" + c.cfg.HostID
+	newPolicyName, err := c.getPolicyName()
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to get new policy name.")
+		return
+	}
+
+	identity, err := c.getAWSIdentity(ctx)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to get AWS identity.")
+		return
+	}
+
+	iamClient, err := c.cfg.Clients.GetAWSIAMClient("")
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to get IAM client.")
+		return
+	}
+
+	findPolicy := func(policyName string) error {
+		switch identity.(type) {
+		case awslib.Role:
+			_, err := iamClient.GetRolePolicyWithContext(ctx, &iam.GetRolePolicyInput{
+				PolicyName: aws.String(policyName),
+				RoleName:   aws.String(identity.GetName()),
+			})
+			return common.ConvertError(err)
+		case awslib.User:
+			_, err := iamClient.GetUserPolicyWithContext(ctx, &iam.GetUserPolicyInput{
+				PolicyName: aws.String(policyName),
+				UserName:   aws.String(identity.GetName()),
+			})
+			return common.ConvertError(err)
+		default:
+			return trace.BadParameter("can only fetch policies for roles or users, got %v", identity)
+		}
+	}
+
+	// Check on old policy. If the old policy does not exist or any other
+	// error, this operation is aborted.
+	if err = findPolicy(oldPolicyName); err != nil {
+		if !trace.IsAccessDenied(err) && !trace.IsNotFound(err) {
+			c.log.WithError(err).Errorf("Failed to get inline policy %v for %v.", oldPolicyName, identity)
+		}
+		return
+	}
+
+	// Wait a bit and check on new policy. Only proceed if new policy is found.
+	c.cfg.Clock.Sleep(10 * time.Minute)
+	if err = findPolicy(newPolicyName); err != nil {
+		c.log.WithError(err).Errorf("Failed to get inline policy %v for %v.", newPolicyName, identity)
+		return
+	}
+
+	// Remove old policy.
+	switch identity.(type) {
+	case awslib.Role:
+		_, err = iamClient.DeleteRolePolicyWithContext(ctx, &iam.DeleteRolePolicyInput{
+			PolicyName: aws.String(oldPolicyName),
+			RoleName:   aws.String(identity.GetName()),
+		})
+	case awslib.User:
+		_, err = iamClient.DeleteUserPolicyWithContext(ctx, &iam.DeleteUserPolicyInput{
+			PolicyName: aws.String(oldPolicyName),
+			UserName:   aws.String(identity.GetName()),
+		})
+	}
+
+	if err != nil && !trace.IsNotFound(common.ConvertError(err)) {
+		c.log.WithError(err).Errorf("Failed to delete inline policy %q for %v. It is recommended to remove this policy since it is no longer required.", oldPolicyName, identity)
```

Comment:
Is there a way to make this more visible? I wouldn't assume that everyone looks into the logs. Could we emit this as an event or something similar that is visible in the UI?
WDYGT?

Reply:
I don't think we have a good way to deliver a message to UI at the moment.  For this particular error, it's not the end of the world. It's actually just our fault couldn't cleanup old policies.